sal_interfaces/ATPtg/ATPtg_Telemetry.xml

```diff
@@ -28,34 +28,13 @@
         <Units>minute</Units>
         <Count>1</Count>
     </item>
-    <item>
-        <EFDB_Name>timeToObservable</EFDB_Name>
-        <Description>Time to current target rising/becoming observable (in minutes)</Description>
-        <IDL_Type>double</IDL_Type>
-        <Units>minute</Units>
-        <Count>1</Count>
-    </item>
     <item>
         <EFDB_Name>timeToUnobservable</EFDB_Name>
         <Description>Time for current target to set/become unobservable (in minutes)</Description>
         <IDL_Type>double</IDL_Type>
         <Units>minute</Units>
```

Comment:
OK, we will change them to absolute timestamps. (TAI seconds, Unix time)
The timeToUnobservable will be changed to

EFDB_Name:timeUnobservable
Description:Time of current target becomes unobservable
IDL_Type:double
Units:second
Count:1

I’ll make the same changes to all the timeTo… items in the currentTimesToLimits. I’ll also change the telemetry topic name to currentTimeOfLimits.

I’ll also change the item ‘timestamp’ to be TAI seconds.

Please let me know if I should make these changes.

Reply:
Not sure I understand your point @pkubanek ... This is a telemetry topic that is streaming constantly. A system that  display this can simply monitor/display the delta time... People are usually more interested in knowing how long it will take to arrive at the limit, not the time it will get there.  A system that monitors/alarms can also do the same, without need to compute delta times. I'd rather keep them as "time until" instead of "time when".

Reply:
My recommendation is to change this. Plus you can have object which will not reach any limit (at least during night, but you might as well have objects close to the pole which will never reach any limit). I solved this on LBTO by publishing NAN as value. Should be documented. Can as well use 0 or negative time value..

You should also publish what caused target to become not observable. It might hit azimuth tracking limit, went to close to pole. LBTO has two derotators, and two sides which might each come with a slightly different time to limit (primary due to different derotators zero offsets).

On LBTO, I come with the following variables (shared memory, similar to what LSST/VRO uses as Telemetry):

```
DOUBLE:PCS/TTL/Side/:normRotHigh:1:degrees:PCS:0:0:0:0:Normalized (0 based) high rotator limit
DOUBLE:PCS/TTL/Side/:normRotLow:1:degrees:PCS:0:0:0:0:Normalized (0 based) low rotator limit
INT:PCS/TTL/Side/:redWarning:1:seconds:PCS:0:0:0:0:Time To Limit yellow warning time (warning will occur this seconds before hitting the limit)
DOUBLE:PCS/TTL/Side/:rotHigh:1:degrees:PCS:0:0:0:0:Time To Limit rotator high limit
DOUBLE:PCS/TTL/Side/:rotLow:1:degrees:PCS:0:0:0:0:Time To Limit rotaror low limit
DOUBLE:PCS/TTL/Side/:timeROT:1:MJDTAI:PCS:0:0:0:0:Rotator time to limit
DOUBLE:PCS/TTL/Side/:timeToLimit:1:MJDTAI:PCS:0:0:0:0:Side time to limit (taking into account Az, El and current rotator)
INT:PCS/TTL/Side/:unwrap:1::PCS:0:0:0:0:Suggested unwrap to increase side rotator TTL
INT:PCS/TTL/Side/:yellowWarning:1:seconds:PCS:0:0:0:0:Time To Limit yellow warning time (warning will occur this seconds before hitting the limit)
INT:PCS/TTL/UnwrapHints/:offset:1:seconds:PCS:0:0:0:0:hint for a minimal time an offset shall remain observable without need to unwrap (without hitting limits)
INT:PCS/TTL/UnwrapHints/:preset:1:seconds:PCS:0:0:0:0:hint for a minimal time a preset shall remain observable without need to unwrap (without hitting limits)
DOUBLE:PCS/TTL/:azHigh:1:degrees:PCS:0:0:0:0:Time To Limits high azimuth limit
DOUBLE:PCS/TTL/:azLow:1:degrees:PCS:0:0:0:0:Time To Limits low azimuth limit
DOUBLE:PCS/TTL/:elHigh:1:degrees:PCS:0:0:0:0:Time To Limits high elevation limit
DOUBLE:PCS/TTL/:elLow:1:degrees:PCS:0:0:0:0:Time To Limit low elevation limit
Side:PCS/TTL/:side:2::PCS:0:0:0:0:
DOUBLE:PCS/TTL/:timeAZ:1:hours:PCS:0.0:0:0:0.0:Warning message of "time to limit" for Azimuth axis
DOUBLE:PCS/TTL/:timeEL:1:hours:PCS:0.0:0:0:0.0:Warning message of "time to limit" for Elevation axis
INT:PCS/TTL/:timeELType:1:N/A:PCS:0:0:0:0:0 - no EL limit, 1 - horizon limit 2 - zenith blind spot
DOUBLE:PCS/TTL/:timeStamp:1:N/A:PCS:0.0:0:0:0.0:TimeStamp when the "time to limit" computation was done (MJD TAI).
INT:PCS/TTL/:unwrap:1::PCS:0:0:0:0:Suggested unwrap to increase azimuth TTL
UnwrapHints:PCS/TTL/:unwrapDefaults:1::PCS:0:0:0:0:
UnwrapHints:PCS/TTL/:unwrapHints:1::PCS:0:0:0:0:
```
how it works (ignore various yellow and red limits):

pcs/TTL/timeStamp is time when calculation was carried
pcs/TTL/timeAZ is NAN on no limit, some value when AZ limit is reached
pcs/TTL/unwrap gives unwraping hints (1 or -1) to reach higher time to limit (e.g. when it's +1, you are expected to turn AZ +360 to reach higher TTL)

Similar for rotators. This helps TCS to properly do unwraps (if needed) when proper button is pushed in PCSGUI.

Reply:
ATPtg_currentTimesToLimits is a separate event. Producing absolute timestamps instead of relative gives you freedom to calculate this once ~30 seconds. It really doesn't change that often, time to limits calculations are relatively CPU intensive (if you use TPK, what TPK does is a stepping to properly calculate limits; that's the only proper way how to track limits hits).

Putting relative timestamps there gives two signals to GUI/consumer:

* you can expect those data often, so you should not update GUI locally. This is wrong. Controller should not be responsible for doing calculations you can handle on consumer. GUI should be updated based on absolute timestamps - you can easily calculate difference and display it in hh:mm:ss format
* to calculate absolute time, you should add timestamps and minutes. You will probably end up doing that on consumer if you drop the first assumption (freqvent updates, which really aren't necessary and will only flood network)

This also points out that rather using telemetry for time to limits (which generally only go down unless you unwrap/change target), it will be better to use event. There isn't need to do often updates, publishing event on changes shall be sufficient.

Reply:
I see, I still think knowing how long it is going to take to arrive at a limit is better and safer than the alternative. We don't have to worry about clock drifts between the clocks on the different services tricking you out. 

As to the reason, note that the same topic will publish the time to each limit in particular, the `timeToUnobservable` is the summary of that.

Reply:
Now this is a good point... We could convert this to an event and have it publish the time it will arrive at the limit and only publish it when it changes. We'll have to talk about this offline though and make a different ticket to make the change.

Reply:
@tribeiro I understand your points. My assumption is you need ~proper time on subscribers anyhow. Other things will go havoc if you don't, and a lot was invested to keep computer clocks synchronized (when your clock drift significantly, https certificates will not work..among other problems). Of course I fully understand the other side and I went through the same discussion when I was designing LBTO interface. I post my decision, which is know to work (ask any LBTO TO) and produce reasons why I believe it's superior to using deltas. You don't expect user to read those values from telemetry, you would like to serve it in a GUI - you can do it and see above for reason why you should remove timestamp + TTL value in minutes on every display node (you will need to do that if you would like to display values going down every GUI update, and don't want to take telemetry bandwidth by publishing time to limits every second; yes, I know this isn't a big deal on local display, but it might be when you for whatever reason would like to have remote display on slow networks).

Reply:
yes.... I can see the benefits of doing it that way... I'll talk to @ayayoshimura about implementing this as an event with that format.

Reply:
Time to limits on tcs-test machine for LBTO takes ~2 seconds from start to end. I cannot confirm what mountain machines gives, as those are shut down, but the mountain machine is same HW as tcs-test. This is for two sides, so it might be reasonable to assume TPK can do calculation for single side in 1 second. Still quite CPU hungry and I don't recommend to run those calculations every second or two.

Reply:
LBTO's TCS machines are 3Ghz 8 core / 16 threads with 128GB RAM.